# Latihan: Data Splitting
<p>Pada kesempatan kali ini Anda akan belajar membagi dataset menggunakan fungsi Train Test Split (holdout method) dari library SKLearn. Ke depannya Anda juga akan mempelajari metode lainnya yang sudah diberi tahu pada sub modul data splitting.</p>
<p>Untuk latihan membagi dataset ini terbilang sangat mudah. Tahapannya terdiri dari langkah-langkah sebagai berikut:</p>
<ol>
    <li>Persiapkan dataset ke dalam Notebook.</li>
    <li>Impor library SKLearn.</li>
    <li>Buat variabel untuk menampung data training dan data testing.</li>
    <li>Panggil fungsi train_test_split().</li>
</ol>
<p>Pada Google Colab, import library yang dibutuhkan.</p>

In [2]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Load dataset:
df = pd.read_csv(r"dataset/train.csv")

# Label encoding untuk kolom kategorical aja
df_lencoder = df.copy()
label_encoder = LabelEncoder()
category_features = df.select_dtypes(include=['object']).columns

#### Handle Missing Values, Label Encoding, Cek Missing Value
Pastikan tidak ada missing values sebelum split

Isi Missing Values yang sifat Numerik dan Kategorikal

In [6]:
# Contoh mengisi nilai yang hilang dengan median untuk kolom numerik
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns
for col in numeric_cols:
    df[col] = df[col].fillna(df[col].median())
    # print(df[col])

categorical_cols = df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

Lalu Label Encoding

In [8]:
df_lencoder = df.copy()
label_encoder = LabelEncoder()
for col in categorical_cols:
    df_lencoder[col] = label_encoder.fit_transform(df[col])

In [9]:
# Cek missing value
print(f"Missing Values: {df_lencoder.isnull().sum().sum()}")

Missing Values: 0


<p>Selanjutnya, Anda harus memiliki data yang akan dibagi. Data ini bisa berupa DataFrame atau array yang berisi fitur (X) dan label/target (y). Pada kasus ini, kita sudah memiliki sebuah dataset yang disimpan pada variabel df_lencoder sehingga Anda hanya perlu memisahkan antara atribut dan label pada dataset tersebut.</p>

In [10]:
for col in category_features:
    df_lencoder[col] = label_encoder.fit_transform(df[col])

# Memisahkan fitur (X) dan target (y)
X = df_lencoder.drop(columns=['SalePrice'])  
y = df_lencoder['SalePrice'] 

print(f"X shape: {X.shape}")
print(f"y shape: {y.shape}")

X shape: (1460, 80)
y shape: (1460,)


<div style="border: 2px solid white">
    <div style="margin: 1%;">
        <p><b>Hasil Anda saat ini:</b> X (1460, 80) artinya 1460 baris, 80 fitur. Ini siap untuk di-split menjadi train-test!</p>
        <p style="margin-bottom: -1.5%;"><b>Arti Dari Fitur:</b></p>
        <p>Fitur adalah kolom/variabel input yang digunakan untuk memprediksi target. Contoh: LotArea, YearBuilt, OverallQual (bukan SalePrice).</p>
        <p><b>Apa yang dimaksud dengan hasil dari y-shape (1460, (dan tidak ada isi disini) )</b></p>
        <p><span style="background-color:gray;">(1460,)</span> artinya 1460 baris, 1 kolom (target <span style="background-color:gray;">SalePrice</span>).</p>
        <p>Tanda koma tanpa angka setelahnya menunjukkan <b>satu dimensi</b> (1D array/series), bukan dua dimensi seperti DataFrame.</p>
    </div>
</div>

<p>Untuk membagi dataset menjadi train set dan test set, kita bisa menggunakan fungsi train_test_split. Fungsi ini memerlukan beberapa parameter utama:</p>
<ul>
    <li style="margin-bottom: 1.5%;"><b>X:</b> kumpulan atribut atau fitur dari dataset, yaitu semua kolom kecuali kolom target (pada kasus ini SalePrice).</li>
    <li style="margin-bottom: 1.5%;"><b>y:</b> target atau label yang akan kita prediksi merupakan kolom yang ingin kita pisahkan (pada kasus ini SalePrice).</li>
    <li style="margin-bottom: 1.5%;"><b>test_size:</b> persentase dari data yang akan digunakan sebagai test set. Misalnya, ketika Anda menentukan test_size=0.2, 20% dari data akan menjadi test set, dan 80% sisanya akan menjadi train set.</li>
    <li style="margin-bottom: 1.5%;"><b>random_state:</b> parameter yang memastikan pemisahan data yang konsisten setiap kali fungsi tersebut dijalankan. Parameter ini mengontrol pengacakan saat membagi dataset sehingga dengan menetapkan random_state hasil pembagian data (train set dan test set) akan selalu sama. Hal ini penting untuk memastikan eksperimen yang dapat direproduksi dan evaluasi model yang konsisten. Namun, parameter ini juga dapat memiliki kekurangan yaitu model tidak akan menerima inputan baru yang mungkin akan lebih baik dari sebelumnya.</li>
</ul>

<p>Ketika Anda menjalankan train_test_split, fungsi ini akan mengembalikan empat bagian data seperti berikut:</p>
<ol>
    <li><b>X_train:</b> fitur untuk train set.</li>
    <li><b>X_test:</b> fitur untuk test set.</li>
    <li><b>y_train:</b> target untuk train set.</li>
    <li><b>y_test:</b> target untuk test set.</li>
</ol>

<p>Ini berarti setelah pemanggilan fungsi, Anda memiliki dua bagian data: satu untuk melatih model (train set) dan satu lagi untuk menguji model (test set).</p>
<p>Bagaimana apakah Anda sudah memahami fondasi data splitting? Bagus, karena seperti biasa teori tanpa implementasi hanyalah sebuah angan-angan. Oleh karena itu, mari kita melangkah sekali lagi untuk melihat contoh kode dari penjelasan di atas:</p>

In [12]:
from sklearn.model_selection import train_test_split
 
# membagi dataset menjadi training dan testing 
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

print(f"x_train: {x_train.shape}")
print(f"x_test: {x_test.shape}")
print(f"y_train: {y_train.shape}")
print(f"y_test: {y_test.shape}")

x_train: (1168, 80)
x_test: (292, 80)
y_train: (1168,)
y_test: (292,)


<b>Keterangan:</b>
<ul>
    <li style="margin-bottom: 1.5%;"><span style="background-color:gray; padding: 0% 1% 0% 1%;">test_size=0.2</span> → 20% data untuk testing (292 baris dari 1460)</li>
    <li style="margin-bottom: 1.5%;"><span style="background-color:gray; padding: 0% 1% 0% 1%;">random_state=1</span> → konsisten tetap running</li>
    <li style="margin-bottom: 1.5%;">80% training (1168 baris), 20% testing (292 baris)</li>
</ul>

<div style="border: 2% solid white;">
    <div style="margin: 1%;">
        <p><b>Pro Tips</b></p>
        <ol>
            <li><b>Selalu Gunakan random_state:</b> untuk memastikan hasil yang konsisten dan dapat direproduksi setiap kali Anda menjalankan kode Anda.</li>
            <li><b>Pertimbangkan Penggunaan stratify:</b> ini sangat penting jika Anda memiliki dataset dengan kelas yang tidak seimbang, karena memastikan bahwa setiap subset memiliki distribusi kelas yang sama.</li>
            <li><b>Evaluasi dengan Cross-Validation:</b> meskipun train_test_split memberikan cara sederhana untuk membagi data, cross-validation dapat memberikan evaluasi model yang lebih robust, terutama jika dataset Anda kecil. (Psst, Anda akan mempelajari metode ini pada materi berikutnya, ya!)</li>
        </ol>
    </div>
</div>

Sangat mudah ‘kan? Tentunya dengan menggunakan metode holdout ini akan mempermudah eksplorasi yang Anda lakukan. Setelah splitting, Anda dapat menggunakan x_train dan y_train untuk melatih model, serta x_test dan y_test untuk menguji kinerjanya. 

In [13]:
# Menghitung panjang / jumlah data
print("Jumlah data: ",len(X))
# Menghitung panjang / jumlah data pada x_train
print("Jumlah data latih: ",len(x_train))
# menghitung panjang/jumlah data pada x_test
print("Jumlah data test: ",len(x_test))

Jumlah data:  1460
Jumlah data latih:  1168
Jumlah data test:  292


Contoh hasil dari Dicoding:<br><br>
![alt text](image/dos-5808419bc25f4c05e92aab5e6ea9bc0420241015125703.jpg)

<p>Sampai di sini data yang Anda miliki sudah siap dilatih pada tahapan pembangunan model yang cukup penting. Oleh karena itu, pastikan Anda sudah melakukan tahapan cleaning sampai splitting dengan benar, ya. Sampai jumpa pada bagian paling seru!</p>